### PR TITLE
Added Image, Snapshot and IP metrics

### PR DESCRIPTION
**Pull request by oswynb**
Added support for Image Events and Pollsters
Added Snapshot Events
Added IP Allocation Events

Updated Infer.hs for instance_disk
Fixed some prism bugs
Added missing tests

@tranma 

Succeeds #3

### CODE REVIEW

**Comment by tranma** [comment]
hold on a sec. I'm cleaning up some of the merged code, then I will rebase this onto master.

**Comment by tranma** [comment]
can you rebase this branch onto master? Use interactive rebase to graft changes & squash commits. Once it's up to date we'll check & merge.

**Comment by tranma** [comment]
Ceilometer.Infer is gone, the code that looks up the schema sourcedict and give you a concrete type is in Ceilometer.Client, you need to add your tags checking there.

File: lib/Ceilometer/Fold.hs [hunk 1]
```diff
@@ -234,4 +292,11 @@ eEvent (start, end) = M.foldlWithKey (\a k v -> a + (fromIntegral k * v)) 0 . go
         go a@_                          = snd a
         s x = max start (x ^. time)
 
+eIPEvent (start, end) = M.foldlWithKey (\a _ v -> a + v) 0 . go . unwrapAcc
+  where -- deal with the dangling last event
+        go (Just x, acc)
+          | d <- end - s x, d > 0 = insertEvent x d acc
+        go a@_                          = snd a
+        s x = max start (x ^. time)
+
```
Comment (tranma) [inline lib/Ceilometer/Fold.hs:297]: this is finicky logic and I'd like to not see it duplicated.

instead, can you change the end event step (eEvent) to take a fold function (acc -> k -> v -> acc) and use eEvent for everything?

File: lib/Ceilometer/Types/Image.hs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+$(makeLenses ''PDImage)
+
+newtype PDImagePollster = PDImagePollster { _pdImagePollsterVal :: PFValue64 }
+     deriving (Show, Read, Eq, Typeable)
+
+pdImagePollster :: Iso' PRSimple PDImagePollster
+pdImagePollster = iso (PDImagePollster . _prSimpleVal) (PRSimple . _pdImagePollsterVal)
+
+pdImage :: Prism' PRCompoundEvent PDImage
+pdImage = prism' pretty parse
```
Comment (tranma) [inline lib/Ceilometer/Types/Image.hs:63]: nitpick: move this up to before newtype PDImagePollster

File: lib/Ceilometer/Tags.hs [hunk 3]
```diff
@@ -46,3 +47,4 @@ lookupMetricName = lookupSource keyMetricName
 lookupEvent      = lookupSource keyEvent
 lookupVolumeType = lookupSource keyVolumeType
 lookupCompound   = lookupSource keyCompound
+isEvent          = maybe True (== valTrue) . lookupEvent
```
Comment (tranma) [inline lib/Ceilometer/Tags.hs]: wait, really? if there's no _event tag, it's considered an event?

Comment (olorin) [inline lib/Ceilometer/Tags.hs]: This seems backwards to me also.

Comment (oswynb) [inline lib/Ceilometer/Tags.hs]: Blindly copied a suggestion from last PR, will fix.

**Comment by tranma** [comment]
remove all the Typeable derivations, they are no longer needed.

Then merge yourself.

**Pull request merged**